### DISCUSSION TITLE
Attritube values are empty

### DISCUSSION THREAD

**Discussion by abt5555** [DotSpatial/DotSpatial#1521] [Q&A]
I want to add point name, easting and northing coordinates of a point as its attribute values. There are related colums but  values are empty. My code is below:

            for (int i = 0; i < featureCollection.Count; i++)
            {
            
                List<MeasuredCoordinate> coordinatestext = new List<MeasuredCoordinate>();
                string outputFilePath = Path.Combine(folderPath, $"R_Line_{100 + i + 1}.txt");
                string textfileName = Path.GetFileNameWithoutExtension(outputFilePath);
               
                using (StreamReader sr = new StreamReader(outputFilePath))
                {
                    string line;
                    while ((line = sr.ReadLine()) != null)
                    {
                        string[] parts = line.Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                        double pn = double.Parse(parts[0]);
                        double x = double.Parse(parts[1]);
                        double y = double.Parse(parts[2]);
                        MeasuredCoordinate coordinate = new MeasuredCoordinate(x, y, pn);
                        coordinatestext.Add(coordinate);
                    }
                }
            
                FeatureSet featureSet = new FeatureSet(FeatureType.Point);
                DataColumn column_pn = new DataColumn("PointID");
                featureSet.DataTable.Columns.Add(column_pn);
                DataColumn column_x = new DataColumn("Easting");
                featureSet.DataTable.Columns.Add(column_x);
                DataColumn column_y = new DataColumn("Northing");
                featureSet.DataTable.Columns.Add(column_y);
                featureSet.Name = textfileName;
            
                foreach (MeasuredCoordinate coordinate in coordinatestext)
                {           
                    NetTopologySuite.Geometries.GeometryFactory geometryFactory = new NetTopologySuite.Geometries.GeometryFactory();
            
                    NetTopologySuite.Geometries.Point textpointPoint = geometryFactory.CreatePoint(coordinate);
                        
                    DataRow newRow = featureSet.DataTable.NewRow();
                    newRow["PointID"] = coordinate.Measure;  
                    newRow["Easting"] = coordinate.X;  
                    newRow["Northing"] = coordinate.Y; 
            
                    featureSet.DataTable.Rows.Add(newRow);

                    featureSet.AddFeature(textpointPoint);
                    var currentMapProjection = form1Instance.map1.Projection;
                    featureSet.Projection = currentMapProjection;
                }
            
                MapPointLayer pointLayer = (MapPointLayer)form1Instance.map1.Layers.Add(featureSet);
                PointSymbolizer symbol = new PointSymbolizer(Color.Blue, DotSpatial.Symbology.PointShape.Ellipse, 3);
                pointLayer.Symbolizer = symbol;
                //form1Instance.map1.Layers.Add(featureSet);
                form1Instance.map1.ZoomToMaxExtent();
                form1Instance.map1.Refresh();
            }

Empty values picture: ![2023-07-27 10_00_41-Window](https://github.com/DotSpatial/DotSpatial/assets/14254710/6059b9f6-fe74-4863-bb63-58e49fe19a6d)

**Comment by jany-tenaj** [comment]
I think you're doing it the wrong way.

Don't add a row to the DataTable but use the resulting Feature from  featureSet.AddFeature(textpointPoint); and fill it's datarow with your values.

**Reply by abt5555** [reply]
Thank you for the response. I updated the code like below and also like you said. It works now. But, only problem "PointID" values are shown as integer although there are 1 decimal numbers.

            for (int i = 0; i < featureCollection.Count; i++)
            {
            
                List<MeasuredCoordinate> coordinatestext = new List<MeasuredCoordinate>();
                string outputFilePath = Path.Combine(folderPath, $"R_Line_{100 + i + 1}.txt");
                string textfileName = Path.GetFileNameWithoutExtension(outputFilePath);
               
                using (StreamReader sr = new StreamReader(outputFilePath))
                {
                    string line;
                    while ((line = sr.ReadLine()) != null)
                    {
                        string[] parts = line.Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                        double pn = double.Parse(parts[0]);
                        double x = double.Parse(parts[1]);
                        double y = double.Parse(parts[2]);
                        MeasuredCoordinate coordinate = new MeasuredCoordinate(x, y, pn);
                        coordinatestext.Add(coordinate);
                    }
                }
            
                IFeatureSet featureSet = new FeatureSet(FeatureType.Point);
                
                featureSet.DataTable.Columns.Add(new DataColumn("PointID", typeof(decimal)));
                featureSet.DataTable.Columns.Add(new DataColumn("Easting", typeof(double)));
                featureSet.DataTable.Columns.Add(new DataColumn("Northing", typeof(double)));
                
                featureSet.Name = textfileName;

                var currentMapProjection = form1Instance.map1.Projection;
                featureSet.Projection = currentMapProjection;

                foreach (MeasuredCoordinate coordinate in coordinatestext)
                {           
                    NetTopologySuite.Geometries.GeometryFactory geometryFactory = new NetTopologySuite.Geometries.GeometryFactory();
            
                    NetTopologySuite.Geometries.Point textpointPoint = geometryFactory.CreatePoint(coordinate);
                        
                    var feat = featureSet.AddFeature(textpointPoint);
                    feat.DataRow["PointID"] = coordinate.Measure;
                    feat.DataRow["Easting"] = coordinate.X;
                    feat.DataRow["Northing"] = coordinate.Y;
                }

                MapPointLayer pointLayer = (MapPointLayer)form1Instance.map1.Layers.Add(featureSet);
                PointSymbolizer symbol = new PointSymbolizer(Color.Blue, DotSpatial.Symbology.PointShape.Ellipse, 3);
                pointLayer.Symbolizer = symbol;

                form1Instance.map1.ZoomToMaxExtent();
                form1Instance.map1.Refresh();
            }

text file format : 1820.0	265947.073	  4598429.008

attribute output picture : 
![2023-07-27 10_56_15-Window](https://github.com/DotSpatial/DotSpatial/assets/14254710/471b3f5a-1202-4c09-84d3-ade1479de4f4)

**Reply by abt5555** [reply]
the  class MeasuredCoordinate :

    public class MeasuredCoordinate : Coordinate
    {
        public double Measure { get; set; }

        public MeasuredCoordinate(double x, double y, double measure)
            : base(x, y)
        {
            Measure = measure;
        }
    }

**Reply by jany-tenaj** [reply]
Why aren't you using double for the following line?

featureSet.DataTable.Columns.Add(new DataColumn("PointID", typeof(decimal)));

 You're using double for all the other places where you use measure.


Your text example contains a value that ends with .0 for your measure value and that corresponds to the value shown in PointID. That PointID doesn't show .0 could simply be caused by the way the numbers are shown. If you want to check whether you can see double values where you expect them you should check values that have values differing from 0 in the decimal places.

**Reply by abt5555** [reply]
featureSet.DataTable.Columns.Add(new DataColumn("PointID", typeof(double))); did not work. That is why I tried decimal instead of double.

Values differing from 0 in the decimal places works. 

example for value 2000.5: 
![2023-07-28 08_48_15-Window](https://github.com/DotSpatial/DotSpatial/assets/14254710/0617f7a1-bbb5-469c-89e9-db59d3a29d98)

The biggest issue for me that the easting and northing values are not correct anymore when I downloaded online map from DotSpatial.Plugins.WebMap because it changes the current map projection. WebMapPlugin.EnableBasemapFetching() makes all layer`s projection as _webMercProj (KnownCoordinateSystems.Projected.World.WebMercator.ToEsriString())

Is there any way to re-project online map projection as the current map projection before opening on the map? Do you have any recommendation or help for this? 

Thank you so much.
